File: java/src/main/java/toolgood/algorithm/internals/functions/datetimes/Function_TIMESTAMP.java
package toolgood.algorithm.internals.functions.datetimes;

import java.util.List;
import java.util.function.BiFunction;

import toolgood.algorithm.AlgorithmEngine;
import toolgood.algorithm.Operand;
import toolgood.algorithm.enums.OperandType;
import toolgood.algorithm.internals.ParameterType;
import toolgood.algorithm.internals.functions.FunctionBase;
import toolgood.algorithm.internals.functions.Function_2;
import toolgood.algorithm.internals.functions.NoneEngine;
import toolgood.algorithm.operands.MyDate;

public final class Function_TIMESTAMP extends Function_2 {
    public Function_TIMESTAMP(FunctionBase[] funcs) {
        super(funcs);
    }

    @Override
    public String Name() {
        return "Timestamp";
    }

    @Override
    public Operand Evaluate(AlgorithmEngine engine, BiFunction<AlgorithmEngine, String, Operand> tempParameter) throws Exception {
        int type = 0;
        if (func2 != null) {
            Operand args2 = GetNumber_2(engine, tempParameter);
            if (args2.IsErrorOrNone()) {
                return args2;
            }
            type = args2.IntValue();
        }

        Operand args0 = GetDate_1(engine, tempParameter);
        if (args0.IsErrorOrNone()) {
            return args0;
        }

        MyDate myDate = args0.DateValue();
        long timestamp = myDate.ToDateTime().getTime();

        if (type == 0) {
            return Operand.Create(timestamp);
        } else if (type == 1) {
            return Operand.Create((double) timestamp / 1000);
        }
        return ParameterError(2);
    }

    @Override
    public OperandType GetResultType() {
        return OperandType.NUMBER;
    }

    @Override
    public void GetParameterTypes(NoneEngine noneEngine, List<ParameterType> result, OperandType operandType,
            String op, String val) {
        func1.GetParameterTypes(noneEngine, result, OperandType.DATE);
        if (func2 != null) {
            func2.GetParameterTypes(noneEngine, result, OperandType.NUMBER);
        }
    }
}

File: csharp/ToolGood.Algorithm/Internals/Functions/DateTimes/Function_TIMESTAMP.cs
using System;
using System.Collections.Generic;
using System.Text;
using ToolGood.Algorithm.Enums;
using ToolGood.Algorithm.Internals;

namespace ToolGood.Algorithm.Internals.Functions.DateTimes
{
	internal sealed class Function_TIMESTAMP : Function_2
	{
		public Function_TIMESTAMP(FunctionBase[] funcs) : base(funcs)
		{
		}

		public override string Name => "Timestamp";

		public override Operand Evaluate(AlgorithmEngine engine, Func<AlgorithmEngine, string, Operand> tempParameter)
		{
			int type = 0; // 类型
			if(func2 != null) {
				var args2 = GetNumber_2(engine, tempParameter);
				if(args2.IsErrorOrNone) { return args2; }
				type = args2.IntValue;
			}
			var args0 = GetDate_1(engine, tempParameter); if(args0.IsErrorOrNone) { return args0; }
			DateTime args1;
			if(engine.UseLocalTime) {
				args1 = args0.DateValue.ToDateTime(DateTimeKind.Local).ToUniversalTime();
			} else {
				args1 = args0.DateValue.ToDateTime(DateTimeKind.Utc);
			}
			if(type == 0) {
				var ms = (args1 - FunctionUtil.StartDateUtc).TotalMilliseconds;
				return Operand.Create(ms);
			} else if(type == 1) {
				var s = (args1 - FunctionUtil.StartDateUtc).TotalSeconds;
				return Operand.Create(s);
			}
			return ParameterError(2);
		}
		public override OperandType GetResultType()
		{
			return OperandType.NUMBER;
		}

		internal override void GetParameterTypes(NoneEngine noneEngine, List<ParameterType> result, OperandType operandType, string op = null, string val = null)
		{
			func1.GetParameterTypes(noneEngine, result, OperandType.DATE);
			if(func2 != null) {
				func2.GetParameterTypes(noneEngine, result, OperandType.NUMBER);
			}
		}
	}

}
